tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -957,112 +957,87 @@ unsigned CGMSHLSLRuntime::ConstructStructAnnotation(
 
   unsigned CBufferSize = CBufferOffset;
 
-  for (RecordDecl::field_iterator Field = RD->field_begin(),
-                                  FieldEnd = RD->field_end();
-       Field != FieldEnd;) {
-    if (Field->isBitField()) {
-      // TODO(?): Consider refactoring, as this branch duplicates much
-      // of the logic of CGRecordLowering::accumulateBitFields().
-      DXASSERT(CGM.getLangOpts().HLSLVersion > hlsl::LangStd::v2015,
-               "We should have already ensured we have no bitfields.");
-      CodeGenTypes &Types = CGM.getTypes();
-      ASTContext &Context = Types.getContext();
-      const ASTRecordLayout &Layout = Context.getASTRecordLayout(RD);
-      const llvm::DataLayout &DataLayout = Types.getDataLayout();
-      RecordDecl::field_iterator End = Field;
-      for (++End; End != FieldEnd && End->isBitField(); ++End)
-        ;
-
-      std::vector<DxilFieldAnnotation> BitFields;
-
-      RecordDecl::field_iterator Run = End;
-      uint64_t StartBitOffset = Layout.getFieldOffset(Field->getFieldIndex());
-      uint64_t Tail = 0;
-      for (; Field != End; ++Field) {
-        uint64_t BitOffset = Layout.getFieldOffset(Field->getFieldIndex());
-        // Zero-width bitfields end runs.
-        if (Field->getBitWidthValue(Context) == 0) {
-          Run = End;
-          continue;
+  // By this point the structType in the StructAnnotation has been stripped of
+  // excess fields if it is a union. As a result we must construct the
+  // annotation differently for unions and structs
+  if (RD->isUnion() && !RD->field_empty()) {
+    FieldDecl *fieldDecl = nullptr;
+    bool largestIsBitfield = false;
+    unsigned largestBitwidth = 0;
+
+    for (FieldDecl *Field : RD->fields()) {
+      CBufferOffset = 0;
+
+      QualType fieldTy = Field->getType();
+
+      CBufferOffset = AlignBaseOffset(fieldTy, CBufferOffset, bDefaultRowMajor,
+                                      CGM, dataLayout);
+
+      const hlsl::ConstantPacking *packOffset = nullptr;
+      for (const hlsl::UnusualAnnotation *it : Field->getUnusualAnnotations()) {
+        switch (it->getKind()) {
+        case hlsl::UnusualAnnotation::UA_ConstantPacking: {
+          packOffset = cast<hlsl::ConstantPacking>(it);
+          CBufferOffset = packOffset->Subcomponent << 2;
+          CBufferOffset += packOffset->ComponentOffset;
+          // Change to byte.
+          CBufferOffset <<= 2;
+        } break;
         }
+      }
 
-        llvm::Type *Type = Types.ConvertTypeForMem(Field->getType());
```

Comment:
I think something just tripped up the diff here. I think this code just all slid lower in the file rather than being removed, but there is a lot of change highlighted here that isn't related to unions.